source/extensions/filters/http/composite/filter.cc

```diff
@@ -0,0 +1,192 @@
+#include "extensions/filters/http/composite/filter.h"
+
+#include "envoy/http/filter.h"
+
+#include "common/common/stl_helpers.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Composite {
+
+namespace {
+// Helper that returns `filter->func(args...)` if the filter is not null, returning `rval`
+// otherwise.
+template <class FilterPtrT, class FuncT, class RValT, class... Args>
+RValT delegateFilterAction(FilterPtrT& filter, FuncT func, RValT rval, Args&&... args) {
```

Comment:
nit: I might call this `delegateFilterActionOr` to match similar things like value_or in optional, etc.